flips/20220203-capability-controllers.md

```diff
@@ -0,0 +1,458 @@
+
+#### Pet names for issued capabilities
+
+If needed the `account.capabilities.issue<T>(path)` can be wrapped into a smart contract function, so that the issuer can more easily keep track of what was issued.
+
+```cadence
+// inside the Counter contract
+access(account) petNames: {String; UInt64}
+
+access(account) issueHasCount(petName: String): Capability<&{HasCount}> {
+   // for brevity this function is not handling pet name collision
+   let cap = self.account.capabilities.issue<&{HasCount}>(/storage/counter)
+   self.petNames[petName] = cap.capabilityID
+   return cap
+}
+```
+
+This allows the account to later access `petNames` when it needs to revoke/retarget a specific capability. This can also be used not just for pet names, but to add other metadata to issued capabilities.
+
+#### Capability Minters
+
+In certain situations it is required that an issuer delegates issuing and revoking capabilities to someone else. In this case the following approach can be used.
+
+Let's assume that the issuer defined an `AdminInterface` resource interface and a `Main` resource (besides the `Counter` and `HasCount` from previous examples).
+
+```cadence
+pub resource interface AdminInterface {
+   fun createCountCap(): Capability<&{HasCount}>
+   fun revokeCountCap(capabilityID: UInt64): Bool
+}
+pub resource Main : AdminInterface {
+   fun createCountCap(): Capability<&{HasCount}> {
+       return self.account.capabilities.issue<&{HasCount}>(/storage/counter)
+   }
+
+   fun revokeCountCap(capabilityID: UInt64) {
+      if let capCon = self.account.capabilities.getController(byCapabilityID: capabilityID) {
+         if capCon.Type != Type<&{HasCount}>() {
+            return false // we have only delegated the issuance/revocation of &{HasCount} capabilities
+         }
+         capCon.revoke()
+      }
+   }
+}
+```
+
+The issuer can then store a `Main` resource in their storage and give the capability to call it to a trusted party. The trusted party can then create and revoke `&{HasCount}` capabilities at will.
+
+```cadence
+issuer.save(<-create Main(), to: /storage/counterMinter)
+let countMinterCap <- issuer.capabilities.issue<&{AdminInterface}>(/storage/counterMinter)
+countMinterCap // give this to a trusted party
+```
+
+It is worth noting that every time the `AdminInterface` is used, a CapCon is created in the issuer's storage, taking up some resources. Therefore revoking capabilities does not free any storage.
```

Comment:
On the note of storage used by CapCons, my understanding is that CapCons remain even after revocation of their associated Capability. Is there a reason why we wouldn't to expose the ability to delete CapCons from the AuthAccount once revoked?

Reply:
That might be a remnant from where it was possible to un-revoke a capability controller. 

Since that is no longer possible there is little reason to keep the CapCon alive after revocation.

Possible reasons would be:
- easier to handle capability logic if CapCon is always present (Capability has to check cap con "Revoked" field before use)
- easier for users to find previously issued capabilities if needed

Maybe its best if CapCons can be deleted and the logic is: 
- if there is no capcon for the capability the capability is considered revoked
- revoking the capcon does not automatically delete it

I definitely agree it would be great to be able to recover the storage space.

Any thoughts?